pkg/controller/master-controller-manager/application-secret-synchronizer/controller.go

```diff
@@ -0,0 +1,172 @@
+	if err != nil {
+		return fmt.Errorf("failed to construct controller: %w", err)
+	}
+
+	if err := c.Watch(&source.Kind{Type: &k8scorev1.Secret{}}, &handler.EnqueueRequestForObject{}, predicate.ByAnnotation(secretTypeAnnotation, "", false)); err != nil {
+		return fmt.Errorf("failed to create watch for secrets: %w", err)
+	}
+
+	return nil
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := r.log.With("resource", request.Name)
+	log.Debug("Processing")
+
+	err := r.reconcile(ctx, log, request)
+	if err != nil {
+		log.Errorw("ReconcilingError", zap.Error(err))
+	}
+
+	return reconcile.Result{}, err
+}
+
+func (r *reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, request reconcile.Request) error {
+	secret := &k8scorev1.Secret{}
+
+	var err error
+	if err = r.masterClient.Get(ctx, request.NamespacedName, secret); err != nil {
+		if !kerrors.IsNotFound(err) {
+			return err
+		} else {
+			// handling deletion
+			delSecret := &k8scorev1.Secret{ObjectMeta: v1.ObjectMeta{Name: request.Name, Namespace: seedNS}}
+			if err := r.handleDeletion(ctx, log, delSecret); err != nil {
+				return fmt.Errorf("failed to delete secret: %w", err)
+			}
+			return nil
+		}
+	}
+
+	seedsecret := secret.DeepCopy()
+	seedsecret.Namespace = seedNS
+	seedsecret.SetResourceVersion("")
+
+	namedSecretCreatorGetter := []reconciling.NamedSecretCreatorGetter{
+		secretCreator(seedsecret),
+	}
+	err = r.reconcileAllSeeds(ctx, log, seedsecret, func(ctx context.Context, c ctrlruntimeclient.Client, o ctrlruntimeclient.Object) error {
+		return reconciling.ReconcileSecrets(ctx, namedSecretCreatorGetter, seedNS, c)
+	})
+	if err != nil {
+		r.recorder.Eventf(secret, k8scorev1.EventTypeWarning, "ReconcilingError", err.Error())
+		return fmt.Errorf("reconciling secret %s failed: %w", seedsecret.Name, err)
+	}
+
+	return nil
+}
+
+func secretCreator(s *k8scorev1.Secret) reconciling.NamedSecretCreatorGetter {
+	return func() (name string, create reconciling.SecretCreator) {
+		return s.Name, func(existing *k8scorev1.Secret) (*k8scorev1.Secret, error) {
+			return s, nil
+		}
+	}
+}
+
+func (r *reconciler) handleDeletion(ctx context.Context, log *zap.SugaredLogger, secret *k8scorev1.Secret) error {
+	delfunc := func(ctx context.Context, c ctrlruntimeclient.Client, o ctrlruntimeclient.Object) error {
+		err := c.Get(ctx, types.NamespacedName{Name: secret.Name, Namespace: secret.Namespace}, &k8scorev1.Secret{})
+		if err != nil {
+			if kerrors.IsNotFound(err) {
+				log.Infof("Secret %q already deleted", secret.Name)
+				return nil
+			}
+			return err
+		}
+
+		return c.Delete(ctx, secret)
+	}
+
+	return r.reconcileAllSeeds(ctx, log, secret, delfunc)
+}
+
+// TODO @vgramer I think this generic version should also work for other controllers like the ApplicationDefinition one. Do you think it makes sense to extract this logic into a central place and also use it for the ADs?
```

Comment:
okay. Yeah so far we only have two, so let's leave it at that. I also realized that it can be easy to add too many arguments to this func. Like possibly some controllers need more information about the seed controller and then you would have to add additional arguments, which are not required by others

⟹ I have removed the comment so far